lib/recursive-methods.rb

```diff
@@ -3,47 +3,62 @@
 # Time complexity: ?
 # Space complexity: ?
 def factorial(n)
-    raise NotImplementedError, "Method not implemented"
+  if n < 0
+    raise ArgumentError
+  elsif (n == 1) || (n== 0)
+    return 1
+  else
+    return n * factorial(n - 1)
+  end
 end
 
 # Time complexity: ?
 # Space complexity: ?
-def reverse(s)
-    raise NotImplementedError, "Method not implemented"
+def reverse(s) 
+  if s.length == 0
+    return s 
+  end
+  a = s[0]
+  b = s.slice(1, s.length - 1)
```

Comment:
s.slice creates a new array and  copies all the individual elements over and so is O(n) by itself.